reaktive/src/commonMain/kotlin/com/badoo/reaktive/observable/SwitchMap.kt

```diff
@@ -0,0 +1,91 @@
+package com.badoo.reaktive.observable
+
+import com.badoo.reaktive.base.ErrorCallback
+import com.badoo.reaktive.base.ValueCallback
+import com.badoo.reaktive.base.subscribeSafe
+import com.badoo.reaktive.base.tryCatch
+import com.badoo.reaktive.disposable.CompositeDisposable
+import com.badoo.reaktive.disposable.Disposable
+import com.badoo.reaktive.disposable.DisposableWrapper
+import com.badoo.reaktive.utils.atomic.AtomicReference
+import com.badoo.reaktive.utils.atomic.update
+import com.badoo.reaktive.utils.atomic.updateAndGet
+
+fun <T, R> Observable<T>.switchMap(mapper: (T) -> Observable<R>): Observable<R> =
+    observable { emitter ->
+        val disposables = CompositeDisposable()
+        emitter.setDisposable(disposables)
+
+        val innerDisposableWrapper = DisposableWrapper()
+        disposables += innerDisposableWrapper
+
+        val state = AtomicReference(SwitchMapState(), true)
+        val serializedEmitter = emitter.serialize()
+
+        subscribeSafe(
+            object : ObservableObserver<T>, ErrorCallback by serializedEmitter {
+                override fun onSubscribe(disposable: Disposable) {
+                    disposables += disposable
+                }
+
+                override fun onNext(value: T) {
+                    serializedEmitter.tryCatch(block = { mapper(value) }, onSuccess = ::onInnerObservable)
+                }
+
+                private fun onInnerObservable(observable: Observable<R>) {
+                    val localDisposableWrapper = DisposableWrapper()
+
+                    /*
+                     * Dispose any existing inner Observable.
+                     * If a previous Observable did not provide its disposable yet it will be disposed automatically later since
+                     * its localDisposableWrapper is disposed.
+                     */
+                    innerDisposableWrapper.set(localDisposableWrapper)
+
+                    val innerObserver =
+                        object : ObservableObserver<R>, ValueCallback<R> by serializedEmitter,
+                            ErrorCallback by serializedEmitter {
+                            override fun onSubscribe(disposable: Disposable) {
+                                localDisposableWrapper.set(disposable)
+                            }
+
+                            override fun onComplete() {
+                                val actualState = state.updateAndGet { previousState ->
```

Comment:
We prefer `if-else` if there are just two possible cases. Also something like `previousState.takeIf { it.innerObserver == this }?.copy(...) ?: previousState` would work.